crates/iota-indexer/src/ingestion/primary/orchestration.rs

```diff
@@ -4,52 +4,111 @@
 
 use std::collections::HashMap;
 
-use iota_data_ingestion_core::WorkerPool;
+use iota_data_ingestion_core::{DataIngestionMetrics, IndexerExecutor, WorkerPool};
 use iota_metrics::get_metrics;
 use iota_types::messages_checkpoint::CheckpointSequenceNumber;
+use prometheus::Registry;
+use tokio::task::JoinHandle;
 use tokio_util::sync::CancellationToken;
 use tracing::info;
 
 use crate::{
-    errors::IndexerError,
-    ingestion::primary::{persist::PrimaryWriter, prepare::PrimaryWorker},
+    ingestion::{
+        common::orchestration::ShimIndexerProgressStore,
+        primary::{persist::PrimaryWriter, prepare::PrimaryWorker},
+    },
     metrics::IndexerMetrics,
-    store::PgIndexerStore,
+    spawn_monitored_task,
+    store::{PgIndexerStore, indexer_store::IndexerStore},
     types::IndexerResult,
 };
+
 const CHECKPOINT_QUEUE_SIZE: usize = 100;
 
-pub async fn setup_primary(
-    state: PgIndexerStore,
-    metrics: IndexerMetrics,
-    checkpoint_download_queue_size: usize,
-) -> Result<(WorkerPool<PrimaryWorker>, PrimaryWriter), IndexerError> {
-    let checkpoint_queue_size = std::env::var("CHECKPOINT_QUEUE_SIZE")
-        .unwrap_or(CHECKPOINT_QUEUE_SIZE.to_string())
-        .parse::<usize>()
-        .unwrap();
-    let global_metrics = get_metrics().unwrap();
-    let (indexed_checkpoint_sender, indexed_checkpoint_receiver) =
-        iota_metrics::metered_channel::channel(
-            checkpoint_queue_size,
-            &global_metrics
-                .channel_inflight
-                .with_label_values(&["checkpoint_indexing"]),
+pub(crate) struct PrimaryPipeline {
+    pub primary_executor: IndexerExecutor<ShimIndexerProgressStore>,
+    pub primary_writer: PrimaryWriter,
+    pub primary_watermark: CheckpointSequenceNumber,
+}
+
+impl PrimaryPipeline {
+    pub async fn setup(
+        state: PgIndexerStore,
+        metrics: IndexerMetrics,
+        checkpoint_download_queue_size: usize,
+        cancel: CancellationToken,
+    ) -> IndexerResult<PrimaryPipeline> {
+        let primary_watermark = state
+            .get_latest_checkpoint_sequence_number()
+            .await
+            .expect("failed to get latest tx checkpoint sequence number from DB")
+            .map(|seq| seq + 1)
+            .unwrap_or_default();
+        let primary_progress_store =
+            ShimIndexerProgressStore::new(vec![("primary".to_string(), primary_watermark)]);
+        let mut primary_executor = IndexerExecutor::new(
+            primary_progress_store,
+            1,
+            DataIngestionMetrics::new(&Registry::new()),
+            cancel.child_token(),
+        );
```

Comment:
Constructing the progress store and the executor is something we do in various places now. So this warrants a new function, e.g. `common::orchestrator::new_executor` to avoid boilerplate, and have single reference point for maintenance.